### PR TITLE
Allow connection timeout to be configured via the environment

### PR DESCRIPTION
**Pull request by mmoayyed**
This pull request allows one to set the connection timeout for tomcat connectors, via a `server.tomcat.connectionTimeout` property. Default is set to -1. 

- [X] I have signed the CLA

### CODE REVIEW

**Comment by philwebb** [comment]
@mmoayyed Is this something you find you commonly need to configure. I'm just wondering if we should add a property or if using a `TomcatConnectorCustomizer` would be enough for most users?

**Comment by mmoayyed** [comment]
Thank you for reviewing this. 

My understanding is that this property allows tomcat to hang on for a while longer until the webapp is a fully ready to be served and has booted up. Presently, with the application I am working with https://github.com/apereo/cas the connection times out which is sort of annoying. When I set out to research this I realized a number of other folks had already chosen the method of a customizer to do which led me to to this pull. I would say it's sort of a common thing and I could probably do away with the customizer and would prefer to do so, but it's really not that pressing.

**Comment by philwebb** [comment]
We'd ideally like to add the same support for Jetty and Undertow (assuming they offer the same feature). Looks like this property eventually maps to `Socket.setSoTimeout`.

**Comment by mmoayyed** [comment]
Sure. Happy to do it. Shall I resume?

**Comment by wilkinsona** [comment]
@mmoayyed Yes please

**Comment by mmoayyed** [comment]
This is ready to be reviewed.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java [hunk 1]
```diff
@@ -1183,6 +1226,17 @@ void customizeUndertow(ServerProperties serverProperties,
...
 			}
+
+			customizeConnectionTimeout(serverProperties, factory);
+		}
+		private void customizeConnectionTimeout(final ServerProperties serverProperties,
+													final UndertowEmbeddedServletContainerFactory factory) {
+			factory.addBuilderCustomizers(new UndertowBuilderCustomizer() {
+				@Override
+				public void customize(Builder builder) {
+					builder.setSocketOption(UndertowOptions.REQUEST_PARSE_TIMEOUT, serverProperties.getConnectionTimeout());
```
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java]: `REQUEST_PARSE_TIMEOUT` is described as "The maximum allowed time of reading HTTP request in milliseconds.". Should this not be `NO_REQUEST_TIMEOUT` which is described as "The amount of time the connection can be idle with no current requests before it is closed"

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java [hunk 2]
```diff
@@ -994,6 +1021,22 @@ void customizeJetty(ServerProperties serverProperties,
...
+		}
+		private void customizeConnectionTimeout(final ServerProperties serverProperties,
+													final JettyEmbeddedServletContainerFactory factory) {
+			factory.addServerCustomizers(new JettyServerCustomizer() {
+				@Override
+				public void customize(final Server server) {
+					for (org.eclipse.jetty.server.Connector connector : server.getConnectors()) {
+						if (connector instanceof ServerConnector) {
+							ServerConnector serverConnector = (ServerConnector) connector;
+							serverConnector.setSoLingerTime(serverProperties.getConnectionTimeout());
```
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java]: The linger time for a socket is very different to the time it'll wait for a connection before being closed. When the socket is closed, the linger time causes the socket to block waiting for acknowledgement of the close from its peer.

Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java]: This should be a call to `setIdleTimeout` instead. It is described as "sets the maximum Idle time for a connection, which roughly translates to the `Socket.setSoTimeout(int)` call.

**Comment by wilkinsona** [comment]
@mmoayyed Thanks again for the PR. Do you have the time to make the two updates that are needed or would you like us to take care of it as part of the merge?

**Comment by mmoayyed** [comment]
I certainly do have the time. I'll probably get the PR updated later today. Thank you very much for the feedback. I'll post here when it's ready for review again.

**Comment by wilkinsona** [comment]
@mmoayyed Thanks once again. I've merged your changes into master along with a [little bit of polish](https://github.com/spring-projects/spring-boot/commit/94ba6bdae26d14cba04fb960fdd6c58b3e789946).

**Pull request closed without merging**